docs/dev/ui_description.rst

```diff
@@ -0,0 +1,264 @@
+up with the state of the widget. That is, we are composing the visual
+representation of the widget, defining the viewer logic, and defining
+its state all during initialization. This is both extremely verbose --
+as the nested nature of the Vuetify library means many of the
+intermediate widget classes (e.g. ``Layout``, ``Container``, ``Row``,
+etc) need to be defined as instance attributes -- and, perhaps more
+severe, makes it more difficult to design reactive UI behavior that
+responds to state changes in the widget classes automatically. This
+means that we must constantly interact with the UI widgets to change
+their state directly instead of simply having the UI *respond* to state
+changes in the custom widget class. Fundamentally, this approach means
+that *in addition to* the state of the custom widget (e.g. ``Toolbar``),
+we must also be aware of the state of *each individual widget* that
+composes it. Put another way, there's no *central source of truth* for
+the state of the custom widget as each element may contain some kind of
+stateful information about itself.
+
+An alternative design, and the one used for the JDAViz tools, is to have each widget implemented as a
+``VuetifyTemplate`` object. In this way, custom widgets are defined as
+ipywidget-like elements whose visual representation is described by a
+Vuetify template. The template composes the visual representation of the
+custom widget using the Vue formalism, while the state is implemented as
+trailets on the custom widget class. The template reads and responds to
+state and state changes on the custom widget, and the custom widget need
+not know or care about how that state is being represented (i.e. the
+only state is that of the custom widget).
+
+With this approach in hand, all widgets in the glupyter GUI are composed
+of two files: the python file declaring the widget class (e.g.
+``Toolbar``), and the ``.vue`` file containing the template describing
+how the widget is to be rendered using components from the Vuetify
+library in the nominal Vue framework. In this way, there is a clear
+separation between the state of each widget (contained in the python
+file) and the view of that state (contained in the Vuetify file). Below
+is an example of what such a widget might looking like:
+
+.. code:: markup
+
+    <template>
+      <v-toolbar>
+        <v-toolbar-side-icon></v-toolbar-side-icon>
+        <v-toolbar-title>Title</v-toolbar-title>
+        <v-spacer></v-spacer>
+        <v-toolbar-items class="hidden-sm-and-down">
+          <v-btn flat>Link One</v-btn>
+          <v-btn flat>Link Two</v-btn>
+          <v-btn flat>Link Three</v-btn>
+        </v-toolbar-items>
+      </v-toolbar>
+    </template>
+
+.. code:: python
+
+    with open(os.path.join(os.path.dirname(__file__), "toolbar.vue")) as f:
+        TEMPLATE = f.read()
+
+
+    class Toolbar(VuetifyTemplate):
+        template = Unicode(TEMPLATE).tag(sync=True)
```

Comment:
Do we really want `template` as a class variable? Maybe....but this means all instances share it, which means we can't customize the Toolbar (e.g.) separately for different instances. That might be the desired behavior, but I worry it could create counter-intuitive behavior, whereas having instance variables with defaults might be more intuitive.

Separately, I'm not wild about using global variables implicitly like `TEMPLATE` in this example. But I'm not sure there is an elegant pythonic way to set the value of `template` in your class definition otherwise. I'm not finding one on quick googling...

Reply:
The class variables are a consequence of the way trailets work, not a design decision we made. Python class variables are unique _per instance_, unless the variable is changed at the class level, which I'm assuming trailets does something special to disallow changing the trailet object instance on class definitions through `setattr`, for example.

> Separately, I'm not wild about using global variables implicitly like TEMPLATE in this example.

We could explore setting the file path in the `__init__` method and having some base class do the parsing and passing. I think currently this is the most straight forward, but we can put some brain power on this.